alg/crc32c_sse42.c

```diff
@@ -50,8 +56,16 @@ CRC32C_Update_SSE42(uint32_t state, const uint8_t * buf, size_t len)
 	 * if we start with i = 4 and want 1 block, we'll end up at i == 12.
 	 */
 	for (; i < in_block; i += 8) {
+#ifdef CPUSUPPORT_X86_SSE42_64
 		state = (uint32_t)_mm_crc32_u64(state,
 		    *(const uint64_t *)&buf[i]);
+#else
+		/* Only process 32 bits at once. */
+		state = (uint32_t)_mm_crc32_u32(state,
+		    *(const uint32_t *)&buf[i]);
+		state = (uint32_t)_mm_crc32_u32(state,
```

Comment:
Is there a reason to not make this a loop with `i += 4` in the 32-bit case?

Reply:
I thought the code was easier to read this way, and the performance impact would be minimal in the 32-bit case.

We could certainly have:

```
#ifdef CPUSUPPORT_X86_SSE42_64
        /*
         * Process blocks of 8.  It's ok if (i % 8 != 0), because we're
         * checking that (i < in_block), not (i != in_block).  For example,
         * if we start with i = 4 and want 1 block, we'll end up at i == 12.
         */
        for (; i < in_block; i += 8) 
                state = (uint32_t)_mm_crc32_u64(state,
                    *(const uint64_t *)&buf[i]);
#else
        /*
         * Process blocks of 4.  It's ok if (i % 4 != 0), because we're
         * checking that (i < in_block), not (i != in_block).  For example,
         * if we start with i = 4 and want 1 block, we'll end up at i == 8.
         */
        for (; i < in_block; i += 4) 
                state = (uint32_t)_mm_crc32_u32(state,
                    *(const uint32_t *)&buf[i]);
#endif
```

but that's also not *ideal* for the 32-bit case: the earlier code ensures that we're processing blocks of 8.  So maybe we should expand that to:
```
#ifdef CPUSUPPORT_X86_SSE42_64
        /* Calculate how many bytes are in aligned blocks. */
        remaining_bytes = len - pre_block;
        in_block = remaining_bytes - (remaining_bytes % 8);

        /* Process bytes before the alignment. */
        for (; i < pre_block; i++)
                state = _mm_crc32_u8(state, buf[i]);

        /* If we would process a block, ensure that it's aligned. */
        assert(!((i < in_block) && !((((uintptr_t)&buf[i]) & 7) == 0)));

        /*
         * Process blocks of 8.  It's ok if (i % 8 != 0), because we're
         * checking that (i < in_block), not (i != in_block).  For example,
         * if we start with i = 4 and want 1 block, we'll end up at i == 12.
         */
        for (; i < in_block; i += 8) 
                state = (uint32_t)_mm_crc32_u64(state,
                    *(const uint64_t *)&buf[i]);

        /* Ensure that we don't have too many bytes remaining. */
        assert((len - i) < 8);
#else
        /* Calculate how many bytes are in aligned blocks. */
        remaining_bytes = len - pre_block;
        in_block = remaining_bytes - (remaining_bytes % 4);

        /* Process bytes before the alignment. */
        for (; i < pre_block; i++)
                state = _mm_crc32_u8(state, buf[i]);

        /* If we would process a block, ensure that it's aligned. */
        assert(!((i < in_block) && !((((uintptr_t)&buf[i]) & 3) == 0)));

        /*
         * Process blocks of 4.  It's ok if (i % 4 != 0), because we're
         * checking that (i < in_block), not (i != in_block).  For example,
         * if we start with i = 4 and want 1 block, we'll end up at i == 8.
         */
        for (; i < in_block; i += 4) 
                state = (uint32_t)_mm_crc32_u32(state,
                    *(const uint32_t *)&buf[i]);

        /* Ensure that we don't have too many bytes remaining. */
        assert((len - i) < 4);
#endif
```

When I was trying to proofread that code, I started to wonder if it was really worth having those extra lines -- particularly because I doubt that anybody will be running this on a 32-bit system.  I figured that moving the `#if` into the `for` loop would allow for a minimal difference between 32-bit and 64-bit, and the slight perfomance impact on 32-bit performance didn't matter.

Reply:
Got it.  Yeah, what you had is fine.  One thing to consider: On 64-bit systems if we need to process e.g. 31 bytes, we could do 3x _mm_crc32_u8, 1x _mm_crc32_u32, and 3x _mm_crc32_u64 rather than using just _u8 and _u64.  The difference won't show up in our testing since we're not using odd lengths though.

Reply:
True... but how complicated do we want `crc32c_sse42.c` to be?  Or rather: do we ever expect CRC32C calculation to be a bottleneck?  I'd expect kivaloo to be limited by network or disk speed.

I mean, if we're benchmarking something on a particular system and this turns out to be an issue, I'm happy to spend more time on it.  But I don't think that we're at that point.

Reply:
Sure.